File: hw6/scala/task2.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.streaming.{Durations, Seconds, StreamingContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, FileOutputStream, PrintWriter}
import java.text.SimpleDateFormat
import java.util.{Calendar, Date}
import scala.math.{abs, pow}
import scala.util.Random
import scala.util.control.Breaks

object task2 {

  def find_prime(x:Int): Int ={
    var isPrime = false
    var x_pri = x
    while (!isPrime) {
      x_pri += 1
      val loop = new Breaks;
      var tmp = 2
      loop.breakable {
        for( i <- 2 until x){
          if (x_pri % i == 0){
            loop.break()
          }
          tmp += 1
        }
      }
      if (tmp == x) {
        isPrime = true
      }
    }
    x_pri
  }

  def generate_hash_parameters(m_min:Int, n_hash_func: Int)= {
    val m = find_prime(m_min)
    val r = Random
    var a_list:List[Long] = List()
    var b_list:List[Long] = List()
    while (a_list.length != n_hash_func) {
      val a = r.nextInt(50000) + 10
      val b = r.nextInt(10000000) + 1000
      if (a % m != 0 && !a_list.contains(a)) {
        a_list = a_list :+ a.toLong
        b_list = b_list :+ b.toLong
      }
    }
    (a_list, b_list, m)
  }

  def my_print(ground_truth: Int, estimation: Int, output_file_path: String): Unit = {
    val now_time = Calendar.getInstance().getTime()
    val t_format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val t = t_format.format(now_time)
    val pw = new PrintWriter(new FileOutputStream(new File(output_file_path), true))
    pw.write(List(t, ground_truth.toString, estimation.toString).mkString(",")+"\n")
    pw.close()
  }

  def predict_num_distinct_value(item: RDD[String], a_list: List[Long], b_list: List[Long], m: Int, n_group: Int, output_file_path: String): Unit = {
    val city_list = item.collect()
    var long_tail_list:List[Int] = List()
    for (i <- a_list.indices) {
      val a = a_list.apply(i)
      val b = b_list.apply(i)
      var long_tail = 0
      for (city <- city_list) {
        val x = abs(city.hashCode).toLong
        val binary_str = ((a * x + b) % m.toLong).toBinaryString
        var tail = 0
        val loop = new Breaks;
        loop.breakable {
          for (e <- binary_str.reverse) {
            if (e == '0') {tail += 1}
            else {loop.break()}
          }
        }
        if (tail > long_tail) {long_tail = tail}
      }
      long_tail_list = long_tail_list :+ long_tail
    }
    val ordered_tail = long_tail_list.sorted
    var group_avg_list:List[Double] = List()
    for (i <- 0 until (ordered_tail.length/n_group).toInt){
      val est = ordered_tail.slice(n_group*i,n_group*(i+1)).map(r=>pow(2, r))
      group_avg_list = group_avg_list :+ est.sum / est.length
    }
    val ordered_avg = group_avg_list.sorted
    val estimation = group_avg_list.apply((ordered_avg.length/2).toInt).toInt
    val ground_truth = city_list.toSet.size
    println(estimation, ground_truth)
    my_print(ground_truth, estimation, output_file_path)
  }

  def main(args: Array[String]): Unit = {
    val start_time = new Date().getTime
    val port = args(0) //"9999"
    val output_file_path = args(1) //"task2_ans_scala.csv"

    val window_length = 30
    val sliding_interval = 10
    val batch_interval  = 5
    val n_hash_func  = 25
    val n_group  = 5
    val hash_params = generate_hash_parameters(10000, n_hash_func)
    val a_list = hash_params._1
    val b_list = hash_params._2
    val m = hash_params._3

    val pw = new PrintWriter(new FileOutputStream(new File(output_file_path), true))
    pw.write(List("Time", "Ground Truth", "Estimation").mkString(",")+"\n")
    pw.close()
    val ssc = new StreamingContext("local[*]", "tasks2", Seconds(batch_interval))
    val sc = ssc.sparkContext
    sc.setLogLevel("OFF")
    val streaming = ssc.socketTextStream("localhost", port.toInt)
    streaming.window(Durations.seconds(window_length), Durations.seconds(sliding_interval))
      .map(item => parse(item)).map(item => compact(item \ "city")).filter(item=>item.trim!="")
      .foreachRDD(item=>predict_num_distinct_value(item, a_list, b_list, m, n_group, output_file_path))

    ssc.start()
    ssc.awaitTermination()
  }
}

File: hw6/python/task2.py
import os
import sys
import csv
import time
import datetime
import json
import random
import binascii
from pyspark import SparkContext
from pyspark.streaming import StreamingContext


def my_print(ground_truth, estimation, output_file_path):
    t = datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d %H:%M:%S')
    with open(output_file_path, 'a') as f:
        f.write(",".join([t, str(ground_truth), str(estimation)])+'\n')
        # writer = csv.writer(f, delimiter=',')
        # print("{},{},{}".format(t, ground_truth, estimation))
        # writer.writerow([t, ground_truth, estimation])


def predict_num_distinct_value(rdd, random_a, random_b, m, n_group, output_file_path):
    city_list = rdd.collect()
    long_tail_list = list()
    for i in range(len(random_a)):
        a, b = random_a[i], random_b[i]
        long_tail = 0
        for city in city_list:
            x = int(binascii.hexlify(city.encode('utf8')), 16)
            hashed_city = (a * x + b) % m
            bit_str = format(hashed_city, 'b')
            tail = 0
            for e in bit_str[::-1]:
                if e == '0': tail += 1
                else: break
            if tail > long_tail:
                long_tail = tail
        long_tail_list.append(long_tail)

    long_tail_list.sort()
    group_avg_list = list()
    for i in range(int(len(long_tail_list)/n_group)):
        est = [2**r for r in long_tail_list[n_group*i:n_group*(i+1)]]
        group_avg_list.append(sum(est)/len(est))
    group_avg_list.sort()
    estimation = int(group_avg_list[int(len(group_avg_list) / 2)])

    ground_truth = len(set(city_list))
    print(estimation, ground_truth)
    my_print(ground_truth, estimation, output_file_path)

def find_prime(x):
    isPrime = False
    while not isPrime:
        x = x + 1
        for i in range(2, x):
            if x % i == 0:
                break
        else:
            isPrime = True
    return x

if __name__ == '__main__':
    os.environ['JAVA_HOME'] = '/home/buyi/app/jdk'
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'
    os.environ['PYSPARK_DRIVER_PYTHON'] = '/usr/bin/python3'

    sc = SparkContext(appName='task1')
    sc.setLogLevel("OFF")
    # port, output_file_path = sys.argv[1:]
    port, output_file_path = "9999", "task2_ans_non_csv"

    window_length, sliding_interval, batch_interval  = 30, 10, 5
    n_hash_func, n_group = 49, 7
    m = find_prime(10000)
    random_a  = random.sample(range(10, 50000), n_hash_func)
    random_b  = random.sample(range(1000, 10000000), n_hash_func)
    with open(output_file_path, 'w') as f:
        f.truncate()
        f.write(",".join(['Time', 'Ground Truth', 'Estimation']) + '\n')
        # writer = csv.writer(f, delimiter=',')
        # writer.writerow(['Time', 'Ground Truth', 'Estimation'])

    ssc = StreamingContext(sc, batch_interval)
    streaming = ssc.socketTextStream('localhost', int(port))

    streaming.window(window_length, sliding_interval).map(json.loads).map(lambda x: x['city'])\
        .filter(lambda x:x.strip()!="").foreachRDD(lambda x:predict_num_distinct_value(x, random_a, random_b, m, n_group, output_file_path))

    ssc.start()
    ssc.awaitTermination()